daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala

```diff
@@ -95,28 +61,52 @@ private[preprocessing] final class TransactionPreprocessor(
       tx: GenTransaction[NodeId, Cid]
   ): (ImmArray[speedy.Command], Set[ContractId]) = {
 
-    type Acc = ((Set[Value.ContractId], Set[Value.ContractId]), BackStack[speedy.Command])
+    type Acc = (Set[Value.ContractId], Set[Value.ContractId], BackStack[speedy.Command])
 
-    val ((localCids, _), cmds) =
-      tx.roots.foldLeft[Acc](((Set.empty, Set.empty), BackStack.empty)) {
-        case ((cids, stack), id) =>
-          tx.nodes.get(id) match {
+    val (_, globalCids, cmds) =
+      tx.roots.foldLeft[Acc]((Set.empty, Set.empty, BackStack.empty)) {
+        case ((localCids0, globalCids0, stack), id) =>
+          val (newGlobals, newLocals, cmd) = tx.nodes.get(id) match {
             case None =>
               fail(s"invalid transaction, root refers to non-existing node $id")
             case Some(node) =>
               node match {
+                case create: Node.NodeCreate[Cid] =>
+                  val (cmd, newCids) =
+                    commandPreprocessor.unsafePreprocessCreate(create.templateId, create.arg)
+                  val newGlobalCids = newCids.filterNot(localCids0)
+                  (newGlobalCids, List(create.coid), cmd)
+                case exe: Node.NodeExercises[_, Cid] =>
+                  val templateId = exe.templateId
+                  val (cmd, newCids) =
+                    commandPreprocessor.unsafePreprocessExercise(
+                      templateId,
+                      exe.targetCoid,
+                      exe.choiceId,
+                      exe.chosenValue,
+                    )
+                  val newGlobalCids = newCids.filterNot(localCids0)
+                  val newLocalCids = GenTransaction(tx.nodes, ImmArray(id)).localContracts.keys
+                  (newGlobalCids, newLocalCids, cmd)
+
                 case _: Node.NodeFetch[_] =>
                   fail(s"Transaction contains a fetch root node $id")
                 case _: Node.NodeLookupByKey[_] =>
                   fail(s"Transaction contains a lookup by key root node $id")
-                case _ =>
-                  val (cmd, acc) = unsafeTranslateNode(cids, node)
-                  (acc, stack :+ cmd)
+                case _: Node.NodeRollback[_] =>
+                  // TODO https://github.com/digital-asset/daml/issues/8020
+                  // how on earth can we turn a rollback node back into a speedy command?
+                  sys.error("rollback nodes are not supported")
               }
           }
+          val globalCids = globalCids0 | newGlobals
+          if (newLocals.exists(globalCids))
+            fail("Conflicting discriminators between a global and local contract ID.")
```

Comment:
Can we get a testcase for this? This seems like a related but not the same bug. I think you can trigger it on a transaction like the following:

1. First do an exercise on a global cid.
2. Then do a create which produces the same discriminator as the global cid.

Reply:
Missed that this just got shuffled around but I still don’t see an existing test for it so it would be nice to add it.

Reply:
It appear the test is not 100% necessary, as the dynamic check performed by the reinterpretation will catch those problematic cases. 
I  still prefer to let the test here. I add a comment and a test.